src/rulesets/Base/indexing.jl

```diff
@@ -48,4 +48,68 @@ function frule((_, ẋ, v̇), ::typeof(setindex!), x::AbstractArray, v, inds...)
 end
 
 
+#####
+##### `eachslice` and friends
+#####
+
+function rrule(::typeof(eachrow), x::AbstractVecOrMat)
+    allrows(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, Val(1)))
+    return collect(eachrow(x)), allrows
+end
+
+function rrule(::typeof(eachcol), x::AbstractVecOrMat)
+    allcols(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, Val(2)))
+    return collect(eachcol(x)), allcols
+end
+
+function rrule(::typeof(eachslice), x::AbstractArray; dims)
+    y = collect(eachslice(x; dims=dims))
+    dim = if length(dims) == 1
+        only(dims)
+    else
+        throw(ArgumentError("""That's amazing, after many years JuliaLang/julia#32310 actually landed.
+            Sadly, the gradient rule for `eachslice` is unable to handle this case right now,
+            please make an issue at https://github.com/JuliaDiff/ChainRules.jl"""))
+    end
+    allslices(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, Val(dim)))
+    return y, allslices
+end
+
+# Using Val(dim) here is worth a factor of 2 in this, on Julia 1.8-
+# @btime rrule(eachcol, $([1 2; 3 4]))[2]($([[10, 20], [30, 40]]))
+function ∇eachslice(dys_raw, x::AbstractArray, vd::Val{dim}) where {dim}
+    dys = unthunk(dys_raw)
+    i1 = findfirst(dy -> dy isa AbstractArray, dys)
+    if i1 === nothing  # all slices are Zero!
+        return _zero_fill!(similar(x, float(eltype(x)), axes(x)))
+    end
+    T = promote_type(eltype(dys[i1]), eltype(x))
+    # The whole point of this gradient is that we can allocate one `dx` array:
+    dx = similar(x, T, axes(x))
+    for i in axes(x, dim)
+        slice = selectdim(dx, dim, i)
+        if dys[i] isa AbstractZero
+            _zero_fill!(slice)  # Avoids this: copyto!([1,2,3], ZeroTangent()) == [0,2,3]
+        else
+            copyto!(slice, dys[i])
+        end
+    end
+    return ProjectTo(x)(dx)
+end
+
+_zero_fill!(dx::AbstractArray{<:Number}) = fill!(dx, zero(eltype(dx)))
+_zero_fill!(dx::AbstractArray) = map!(zero, dx, dx)
+
+function rrule(::typeof(∇eachslice), dys, x, vd::Val)
+    function ∇∇eachslice(dz_raw)
+        dz = unthunk(dz_raw)
+        iter = vd == Val(1) ? eachrow(dz) : vd == Val(2) ? eachcol(dz) : eachslice(dz; dims=_val(vd))
```

Comment:
Why is this not just:
```suggestion
        iter = eachslice(dz; dims=_val(vd))
```

is constant folding breaking?

Reply:
I don't think eachslice behaves well, e.g.:
```
julia> @code_warntype eachcol(rand(3,3))
MethodInstance for eachcol(::Matrix{Float64})
  from eachcol(A::AbstractVecOrMat) in Base at abstractarraymath.jl:583
Arguments
  #self#::Core.Const(eachcol)
  A::Matrix{Float64}
Locals
  #243::Base.var"#243#244"{Matrix{Float64}}
Body::Base.Generator{Base.OneTo{Int64}, Base.var"#243#244"{Matrix{Float64}}}

julia> _val(::Val{X}) where X = X
_val (generic function with 1 method)

julia> @code_warntype eachslice(rand(3,3); dims=_val(Val(1)))
MethodInstance for (::Base.var"#eachslice##kw")(::NamedTuple{(:dims,), Tuple{Int64}}, ::typeof(eachslice), ::Matrix{Float64})
  from (::Base.var"#eachslice##kw")(::Any, ::typeof(eachslice), A::AbstractArray) in Base at abstractarraymath.jl:621
Arguments
  _::Core.Const(Base.var"#eachslice##kw"())
  @_2::NamedTuple{(:dims,), Tuple{Int64}}
  @_3::Core.Const(eachslice)
  A::Matrix{Float64}
Locals
  dims::Int64
  @_6::Int64
Body::Base.Generator{Base.OneTo{Int64}}  # bright red

julia> _es(x, v) = eachslice(x; dims=_val(v))
_es (generic function with 1 method)

julia> @code_warntype _es(rand(3,3), Val(1))
MethodInstance for _es(::Matrix{Float64}, ::Val{1})
  from _es(x, v) in Main at REPL[22]:1
Arguments
  #self#::Core.Const(_es)
  x::Matrix{Float64}
  v::Core.Const(Val{1}())
Body::Base.Generator{Base.OneTo{Int64}}  # the same
```

Reply:
```suggestion
        # eachslice(dz; dims=_val(vd)) does not make @code_warntype happy
        iter = vd == Val(1) ? eachrow(dz) : vd == Val(2) ? eachcol(dz) : eachslice(dz; dims=_val(vd))
```